src/experimental/tools/VideoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,306 @@
+   * - Append data
+   * - Set time on video element
+   * @param {Object} contentInfos
+   * @param {Object} contentInfos
+   * @param {Object} contentInfos
+   * @returns {Promise}
+   */
+  private _startTimeSettingTask(contentInfos: IContentInfos,
+                                segments: ISegment[],
+                                time: number
+  ): Promise<number> {
+    const loader = this._loaders[contentInfos.manifest.transport];
+    if (loader === undefined) {
+      const error =
+        new VideoThumbnailLoaderError("NO_LOADER",
+                                      "VideoThumbnailLoaderError: No " +
+                                      "imported loader for this transport type: " +
+                                      contentInfos.manifest.transport);
+      return PPromise.reject(error);
+    }
+    const killTask$ = new Subject();
+    const abortError$ = killTask$.pipe(
+      map(() => {
+        throw new VideoThumbnailLoaderError("ABORTED",
+                                            "VideoThumbnailLoaderError: Aborted job.");
+      })
+    );
+
+    const segmentLoader = createSegmentLoader(
+      loader.video.loader,
+      undefined,
+      { baseDelay: 0,
+        maxDelay: 0,
+        maxRetryOffline: 0,
+        maxRetryRegular: 0 }
+    );
+    const { parser: segmentParser } = loader.video;
+
+    const taskPromise = observableRace(
+      abortError$,
+      getInitializedSourceBuffer$(contentInfos,
+                                  this._videoElement,
+                                  { segmentLoader,
+                                    segmentParser }).pipe(
+        mergeMap((videoSourceBuffer) => {
+          const bufferCleaning$ = removeBufferAroundTime$(this._videoElement,
+                                                          videoSourceBuffer,
+                                                          time);
+          log.debug("VTL: Removing buffer around time.", time);
+
+          const segmentsLoading$ =
+            loadSegments(segments, this._requests, segmentLoader, contentInfos);
+
+          return observableMerge(bufferCleaning$.pipe(ignoreElements()),
+                                 segmentsLoading$
+          ).pipe(
+            mergeMap((arr) => {
+              return combineLatest(
+                arr.map(({ segment, data }) => {
+                  const start = segment.time / segment.timescale;
+                  const end = start + (segment.duration / segment.timescale);
+                  const inventoryInfos = objectAssign({ segment,
+                                                        start,
+                                                        end }, contentInfos);
+                  return pushData(inventoryInfos,
+                                  segmentParser,
+                                  data,
+                                  videoSourceBuffer)
+                    .pipe(tap(() => {
+                      this._requests.delete(segment.id);
```

Comment:
I would think it's more `loadSegments`'s role to delete those, no? It's that function whose adding it.

Maybe even making it a class (or function returning another function) which would have the internal _requests state (or why not even a global?)